packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+  // we can generalize this when needed later
+  @Throws(IOException::class)
+  private fun ensurePermission(uri: Uri, permission: Permission, errorMsg: String) {
+    if (permissionsForUri(uri)?.contains(permission) != true) {
+      throw IOException(errorMsg)
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun ensurePermission(uri: Uri, permission: Permission) {
+    if (permission == Permission.READ) {
+      ensurePermission(uri, permission, "Location '$uri' isn't readable.")
+    }
+    if (permission == Permission.WRITE) {
+      ensurePermission(uri, permission, "Location '$uri' isn't writable.")
+    }
+    ensurePermission(uri, permission, "Location '" + uri + "' doesn't have permission '" + permission.name + "'.")
+  }
+
+  @Throws(IOException::class)
+  private fun openAssetInputStream(uri: Uri): InputStream {
+    // AssetManager expects no leading slash.
+    val asset = uri.path!!.substring(1)
+    return context.assets.open(asset)
+  }
+
+  @Throws(IOException::class)
+  private fun openResourceInputStream(resourceName: String?): InputStream {
+    var resourceId = context.resources.getIdentifier(resourceName, "raw", context.packageName)
+    if (resourceId == 0) {
+      // this resource doesn't exist in the raw folder, so try drawable
+      resourceId = context.resources.getIdentifier(resourceName, "drawable", context.packageName)
+      if (resourceId == 0) {
+        throw FileNotFoundException("No resource found with the name $resourceName")
+      }
+    }
+    return context.resources.openRawResource(resourceId)
+  }
+
+  @ExpoMethod
+  fun getInfoAsync(_uriStr: String, options: Map<String?, Any?>, promise: Promise) {
+    var uriStr = _uriStr
+    try {
+      val uri = Uri.parse(uriStr)
+      var absoluteUri = uri
+      if (uri.scheme == "file") {
+        uriStr = uriStr.substring(uriStr.indexOf(':') + 3)
+        absoluteUri = Uri.parse(uriStr)
+      }
+      ensurePermission(absoluteUri, Permission.READ)
+      if (uri.scheme == "file") {
+        val file = uriToFile(absoluteUri)
+        if (file.exists()) {
+          promise.resolve(
+            Bundle().apply {
+              putBoolean("exists", true)
+              putBoolean("isDirectory", file.isDirectory)
+              putString("uri", Uri.fromFile(file).toString())
+              putDouble("size", getFileSize(file).toDouble())
+              putDouble("modificationTime", 0.001 * file.lastModified())
+              if (options.containsKey("md5") && (options["md5"] == true)) {
+                putString("md5", md5(file))
+              }
+            }
+          )
+        } else {
+          promise.resolve(
+            Bundle().apply {
+              putBoolean("exists", false)
+              putBoolean("isDirectory", false)
+            }
+          )
+        }
+      } else if (uri.scheme == "content" || uri.scheme == "asset" || uri.scheme == null) {
+        try {
+          val `is`: InputStream = when (uri.scheme) {
```

Comment:
```suggestion
          val inputStream: InputStream = when (uri.scheme) {
```
I'd rename this to avoid the backticks